homeassistant/components/ws66i/media_player.py

```diff
@@ -0,0 +1,199 @@
+
+    return [source_id_name, source_name_id, source_names]
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the WS66i 6-zone amplifier platform."""
+    ip_addr = config_entry.data[CONF_IP_ADDRESS]
+    ws66i = hass.data[DOMAIN][config_entry.entry_id][WS66I_OBJECT]
+    sources = _get_sources_from_dict(config_entry.options)
+
+    # Build the entities that control each of the zones.
+    # Zones 11 - 16 are the master amp
+    # Zones 21,31 - 26,36 are the daisy-chained amps
+    entities = []
+    for amp_num in range(1, 4):
+
+        if amp_num > 1:
+            # Don't add entities that aren't present
+            status = await hass.async_add_executor_job(
+                ws66i.zone_status, (amp_num * 10 + 1)
+            )
+            if status is None:
+                break
+
+        _LOGGER.info("Detected amp %d at ip %s", amp_num, ip_addr)
+        for zone_num in range(1, 7):
+            zone_id = (amp_num * 10) + zone_num
+            entities.append(Ws66iZone(ws66i, sources, config_entry.entry_id, zone_id))
+
+    async_add_entities(entities)
+
+    # Set up services
+    platform = entity_platform.async_get_current_platform()
+
+    platform.async_register_entity_service(
+        SERVICE_SNAPSHOT,
+        {},
+        "snapshot",
+    )
+
+    platform.async_register_entity_service(
+        SERVICE_RESTORE,
+        {},
+        "restore",
+    )
+
+
+class Ws66iZone(MediaPlayerEntity):
+    """Representation of a WS66i amplifier zone."""
+
+    def __init__(self, ws66i, sources, namespace, zone_id):
+        """Initialize new zone."""
+        self._ws66i = ws66i
+        self._zone_id = zone_id
+        # dict source_id -> source name
+        self._source_id_name = sources[0]
+        # dict source name -> source_id
+        self._source_name_id = sources[1]
+        # ordered list of all source names
+        self._attr_source_list = sources[2]
+        self._attr_unique_id = f"{namespace}_{self._zone_id}"
+        self._attr_name = f"Zone {self._zone_id}"
+        self._attr_supported_features = SUPPORT_WS66I
+        self._attr_available = True
+        self._attr_should_poll = True
+        self._attr_is_volume_muted = None
+        self._attr_source = None
+        self._attr_state = None
+        self._attr_media_title = None
+        self._attr_device_info = {
+            "identifiers": {(DOMAIN, self.unique_id)},
+            "name": self.name,
+            "manufacturer": "Soundavo",
+            "model": "WS66i 6-Zone Amplifier",
+        }
+
+        self._snapshot = None
+        self._volume = None
+
+    def update(self):
+        """Retrieve latest state."""
+        new_state = self._ws66i.zone_status(self._zone_id)
+        if not new_state:
+            # End-user should turn on Amp and reload integration
+            _LOGGER.warning("Zone %d not detected. Is it turned off?", self._zone_id)
```

Comment:
https://developers.home-assistant.io/docs/integration_quality_scale_index/#silver-

Handles device/service unavailable. Log a warning once when unavailable, log once when reconnected.

You can avoid all the complexity of handling that by using the coordinator https://developers.home-assistant.io/docs/integration_fetching_data?_highlight=datau#coordinated-single-api-poll-for-data-for-all-entities

Reply:
I don't understand how coordinators work. The HA documentation doesn't go into detail how it works, so I have to look at other integration's examples to figure it out. I make an observation in an example for instance, and compare it with the others. If I see a pattern among them, I try to understand why they are doing it the way they are. Most seem to create the DataUpdateCoordinator in the __init__.py file and the CoordinatorEntity in the platform file. Not everyone follows this convention, but I assume it's the preferred way to organize things. There is always 1 DataUpdateCorrdinator and 1 or more CoordinatorEntitys. I need at least 6 entities to represent each of the zones on the WS66i, so I assume I want to create 6 of them. But I don't understand how those 6 entities get updated. 

Before coordinators, I would create my 6 entities and call `async_add_entities()` in my platform file. HA uses this probably to iterate through all entities with polling enabled, calls it's `update()` method, and then reads each of the entities attributes. This makes sense; I understand how HA retrieves my entities data... But with coordinators, it's not clear exactly what it does. It looks like I still add my entities (but now which inherits from CoordinatorEntity) through the `async_add_entities() method`. So, do each of my entities `update()` method still get called? If so, then what is the point of the DataUpdateCoordinator, which the documentation states helps me update all my entities from a single endpoint. In fact, I see from the example in the docs that a coordinator is created, then `async_config_entry_first_refresh()` is called before any of the entities are created! So how do the entities get updated from this call?!

I've tried to rationalize that HA will call my coordinator's `_async_update_data()`, but I don't see how HA knows to call it. But if it somehow knows and does call it, then what I would think it would do it look at each entity that is manages, then call each of those entities `update()` methods. Yet this is what I don't see. It would make sense if it worked this way, but that is not the case. In fact by looking at the other integrations examples, I don't see how the coordinator even is aware of my entities, so I am still confused how they get updated. It only looks like my entities are aware of the coordinator, not the other way around.

What is `coordinator.data`? I see the example in the docs uses it when creating entities, but does not explain why it is doing that or what it is used for... What methods am I expected to implement? What methods does HA call, and in what order?

Sorry I am just ranting right now. I feel like the documentation needs a lot of work, not just the coordinator section. And I'm not alone with this feeling. There was a whole reddit post about it https://www.reddit.com/r/homeassistant/comments/svg8s8/home_assistant_docs_is_a_nightmare_a_new_user/ .

Anyway, I just need to take a break and try to understand this after some rest.

Reply:
You could move the zone discovery into `__init__.py` into a new function along with `open` so if it does fail then setup would be retried. Then store the results along side the data in `hass.data.setdefault(DOMAIN, {})[entry.entry_id]`  and then create a coordinator per zone (example https://github.com/home-assistant/core/blob/dev/homeassistant/components/lookin/__init__.py#L114)

If you get stuck, I can work up an example for you once the other comments are addressed

Reply:
I managed to figure it out. Thanks for the offer.